apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl

```diff
@@ -0,0 +1,473 @@
+    {ok, [], St};
+%%
+%% Connecting
+%%
+on_info(St0, ?leader_connect_response_match(Leader)) when ?is_connecting(St0) ->
+    St1 = St0#{status => ?connected, leader => Leader},
+    St2 = cancel_timer(St1, ?find_leader_timer),
+    St3 = ensure_timer(St2, ?ping_leader_timer),
+    {ok, [], St3};
+on_info(#{id := Id, share_topic_filter := ShareTopicFilter} = St, ?find_leader_timer) when
+    ?is_connecting(St)
+->
+    ?tp(debug, ds_shared_sub_ssubscriber_find_leader_timeout, #{
+        ssubscriber_id => ?format_ssubscriber_id(Id),
+        share_topic_filter => ShareTopicFilter
+    }),
+    ok = emqx_ds_shared_sub_registry:leader_wanted(Id, ShareTopicFilter),
+    {ok, [], ensure_timer(St, ?find_leader_timer)};
+%%
+%% Ping leader, status independent
+%%
+on_info(St0, ?leader_ping_response_match(_)) ->
+    St1 = cancel_timer(St0, ?ping_leader_timeout_timer),
+    St2 = ensure_timer(St1, ?ping_leader_timer),
+    {ok, [], St2};
+on_info(#{session_id := SessionId, id := Id} = St, ?ping_leader_timer) ->
+    ?tp(debug, ds_shared_sub_ssubscriber_ping_leader, #{
+        session_id => SessionId,
+        ssubscriber_id => ?format_ssubscriber_id(Id)
+    }),
+    ok = notify_ping(St),
+    {ok, [], ensure_timer(St, ?ping_leader_timeout_timer)};
+on_info(#{session_id := SessionId, id := Id} = St, ?ping_leader_timeout_timer) ->
+    ?tp(warning, ds_shared_sub_ssubscriber_ping_leader_timeout, #{
+        session_id => SessionId,
+        ssubscriber_id => ?format_ssubscriber_id(Id)
+    }),
+    reset(St, revoke_all_events(St));
+%%
+%% Grant stream
+%%
+on_info(St0, ?leader_grant_match(_Leader, _StreamProgress)) when ?is_unsubscribing(St0) ->
+    %% If unsubscribing, ignore the grant
+    {ok, [], St0};
+on_info(St0, ?leader_grant_match(Leader, _StreamProgress) = Msg) when ?is_connecting(St0) ->
+    St1 = St0#{leader => Leader, status => ?connected},
+    St2 = cancel_timer(St1, ?find_leader_timer),
+    St3 = ensure_timer(St2, ?ping_leader_timer),
+    on_info(St3, Msg);
+on_info(
+    #{streams := Streams0, session_id := SessionId, id := Id} = St0,
+    ?leader_grant_match(_Leader, #{stream := Stream, progress := Progress0})
+) when ?is_connected(St0) ->
+    ?tp(debug, ds_shared_sub_ssubscriber_leader_grant, #{
+        session_id => SessionId,
+        ssubscriber_id => ?format_ssubscriber_id(Id)
+    }),
+    {Events, Streams1} =
+        case Streams0 of
+            #{Stream := _} ->
+                {[], Streams0};
+            #{} ->
+                {
+                    [lease_event(Stream, Progress0)],
+                    Streams0#{Stream => new_stream_data(Progress0)}
+                }
+        end,
+    St1 = St0#{streams => Streams1},
+    ok = notify_progress(St1, Stream),
+    {ok, Events, St1};
+%%
+%% Revoke stream
+%%
+on_info(St, ?leader_revoke_match(_Leader, _Stream)) when ?is_connecting(St) ->
+    %% Should never happen.
+    reset(St);
```

Comment:
Nit. Could be great to have better visibility into such events. Or perhaps something like a test-profile-only assertion?